### PR TITLE
Always use inf_batch_size during inference, even without retrieval

### PR DESCRIPTION
**Pull request by j201**
Ensures we don't OOM on very large test sets. Downside is that it'll slow down inference when retrieval isn't used.

Also use inf_batch_size on CPU when explicitly specified, and warn user to set it explicitly instead of silently overriding with a really low and slow value (this has made it look unexpectedly slow for me when testing on CPU before).

Also get rid of the last `squeeze()` without a dim specified.

In a future update, we should do the following breaking change, since memory usage is `O(context_size + inf_batch_size)` without retrieval but `O(context_size * inf_batch_size)` with retrieval:
- Move inf_batch_size to be a predict parameter
- If `None`, set the default based on both the device and whether retrieval is used, suggested values:
    - GPU+No retrieval: 1M
    - GPU+Retrieval: 512
    - CPU+No Retrieval:  32k
    - CPU+Retrieval: 16

Probably only needs 1 review, whoever's free.

### CODE REVIEW

File: src/tabdpt/classifier.py [hunk 1]
```diff
@@ -89,26 +89,30 @@ def predict_proba(
...
 
-            if self.num_classes <= self.max_num_classes:
-                pred = self.model(
-                    x_src=torch.cat([X_train, X_test], dim=1),
-                    y_src=y_train.unsqueeze(-1),
-                    task=self.mode,
-                )
-            else:
-                pred = self._predict_large_cls(X_train, X_test, y_train)
+            for b in range(math.ceil(len(self.X_test) / self.inf_batch_size)):
```
Comment (anthonycaterini) [inline src/tabdpt/classifier.py:98]: Given this change, I suggest combining the retrieval and non-retrieval branches of the code. A lot of the stuff is the same.

File: src/tabdpt/classifier.py [hunk 2]
```diff
@@ -89,26 +89,30 @@ def predict_proba(
...
-            pred_val = pred.float().squeeze().detach().cpu().numpy()
+                if self.num_classes <= self.max_num_classes:
+                    pred = self.model(
+                        x_src=torch.cat([X_train, X_test[:, start:end]], dim=1),
+                        y_src=y_train.unsqueeze(-1),
+                        task=self.mode,
+                    )
+                else:
+                    pred = self._predict_large_cls(X_train, X_test[:, start:end], y_train)
+
```
Comment (anthonycaterini) [inline src/tabdpt/classifier.py:110]: why do we not do `pred.float()` here but we do on line 142?

File: src/tabdpt/classifier.py [hunk 3]
```diff
@@ -142,7 +146,8 @@ def predict_proba(
                     pred /= pred.sum(axis=-1, keepdims=True)  # numerical stability
 
                 pred_list.append(pred.squeeze(dim=0))
-            pred_val = torch.cat(pred_list, dim=0).squeeze().detach().cpu().float().numpy()
+
+        pred_val = torch.cat(pred_list, dim=0).detach().cpu().float().numpy()
```
Comment (anthonycaterini) [inline src/tabdpt/classifier.py:150]: what is the dimensionality of each element of `pred_list`? How are we able to remove the `squeeze` ?

File: src/tabdpt/classifier.py [hunk 4]
```diff
@@ -13,7 +13,7 @@
 class TabDPTClassifier(TabDPTEstimator, ClassifierMixin):
     def __init__(
         self,
-        inf_batch_size: int = 512,
+        inf_batch_size: int = None,
```
Comment (anthonycaterini) [inline src/tabdpt/classifier.py]: we should consider removing the default argument for `inf_batch_size` (not setting as `None`) for `TabDPTClassifier`

File: src/tabdpt/estimator.py [hunk 5]
```diff
@@ -33,7 +33,7 @@ def download_weights() -> str:
     def __init__(
         self,
         mode: Literal["cls", "reg"],
-        inf_batch_size: int = 512,
+        inf_batch_size: int = None,
```
Comment (anthonycaterini) [inline src/tabdpt/estimator.py]: technically the type hint now should be `inf_batch_size: int | None = None`

File: src/tabdpt/regressor.py [hunk 6]
```diff
@@ -12,7 +12,7 @@
 class TabDPTRegressor(TabDPTEstimator, RegressorMixin):
     def __init__(
         self,
-        inf_batch_size: int = 512,
+        inf_batch_size: int = None,
```
Comment (anthonycaterini) [inline src/tabdpt/regressor.py]: same comment about removing default argument

File: src/tabdpt/classifier.py (hunk 4 above)
Comment (anthonycaterini) [reply src/tabdpt/classifier.py]: If it's a backwards compatibility thing, then we should consider adding a warning when there is no argument present?

File: src/tabdpt/regressor.py [hunk 7]
```diff
@@ -52,19 +52,25 @@ def _predict(self, X: np.ndarray, context_size: int | None = 2048, seed: int | N
...
             X_test = pad_x(test_x[None, :, :], self.max_features).to(self.device)
             y_train = train_y[None, :].float()
-            pred = self.model(
-                x_src=torch.cat([X_train, X_test], dim=1),
-                y_src=y_train.unsqueeze(-1),
-                task=self.mode,
-            )
 
-            return pred.float().squeeze(1).detach().cpu().float().numpy()
+            for b in range(math.ceil(len(self.X_test) / self.inf_batch_size)):
```
Comment (anthonycaterini) [inline src/tabdpt/regressor.py:61]: same comment about DRY implementation of this with the similar loop for retrieval

**Review by anthonycaterini — CHANGES_REQUESTED** [review]
Some comments given. No real concern about functionality but suggest changing some minor things before merging

File: src/tabdpt/classifier.py (hunk 1 above)
Comment (j201) [reply src/tabdpt/classifier.py:98]: I looked at doing that but there are a number of minor differences that make it awkward (e.g., how you slice test samples is different because the non-retrieval branch adds a dummy dimension early on, the retrieval branch does additional normalization post-softmax that I don't really understand)

File: src/tabdpt/classifier.py (hunk 2 above)
Comment (j201) [reply src/tabdpt/classifier.py:110]: Beats me, I think the legacy code had a bunch of extraneous, paranoid `.float()` and `.squeeze()` calls, but it's a bit tough to tell which ones are because of subtle differences

File: src/tabdpt/classifier.py (hunk 3 above)
Comment (j201) [reply src/tabdpt/classifier.py:150]: `n_test x n_class`
In the retrieval branch, there's an extra dimension of size one in position 0 because we batch in the second dimension but only have one eval sample for each prediction in the first dimension, and that gets removed at line 144. In the non-retrieval branch, the extra dimension is instead in position 1 because we don't do any batching in the second dimension and put all the eval samples in the first dimension, and that gets removed at line 114. We shouldn't be doing any squeezing after that, it got left in as a noop (verified stepping through each branch with `tests/cls_example.py`).

I think this actually might make the code at lines 153-154 redundant, but it doesn't hurt to leave it.

File: src/tabdpt/estimator.py (hunk 5 above)
Comment (j201) [reply src/tabdpt/estimator.py]: Fixed

File: src/tabdpt/classifier.py (hunk 4 above)
Comment (j201) [reply src/tabdpt/classifier.py]: Do we want to require users to specify it though? Not sure that's necessary

Comment (anthonycaterini) [reply src/tabdpt/classifier.py]: Sorry, I had the inheritance / workflow backwards in my head, thinking that estimator was calling regressor or classifier, rather than the other way around. this is fine

**Review by anthonycaterini — APPROVED** [review]
Looks good now, agreed with all the feedback (+ my bad for the default argument comment)
